One stripslashes() too many, and some misplaced.

adam-vessey opened the pull request:
**JIRA Ticket**: [ISLANDORA-1889](https://jira.duraspace.org/browse/ISLANDORA-1889)

Supercedes #311.

# What does this Pull Request do?

Make display of query input consistent.

# What's new?

1. Removal of a `stripslashes()` invocation.
2. Addition of forward slash to be escaped to support newer Solr versions.
3. Fix use of `stripslashes()` in advanced searches with multiple terms.

# How should this be tested?

1. Ingest objects with a variety of forward and backslashes and spaces in their titles.
2. [Configure](https://github.com/Islandora/islandora_solr_facet_pages/pull/32#issuecomment-283071291) to use the new regex escape functionality. 
3. Perform simple and advanced searches for these objects using the exact titles.

# Interested parties

@DiegoPino @MorganDawe @willtp87

adam-vessey committed:
One stripslashes() too many.

willtp87 commented:
I was unable to break this change.

adam-vessey committed:
Add forward slash to regex for escaping.

Solr 4.0+ introduced regex search using forward slash as a delimiter, so
needs to be escaped. For backwards compat (to 3.6.2 at least), escaped
forward slashes will be handled as normal forward slashes.

DiegoPino commented:
@willtp87 @adam-vessey do you have a minute here?
The new regex/less stripslashes solved various issues:
  * Solr 400 errors are gone (because slash is in the regex)
  * Breadcrumbs and search field value is kept.

But this whole escaping thing is really messy (sorry) and not sure how many iterations we need for this:

I added a new object with label `//Object of affection` 😁 

Well, I assumed advanced search considers that the whole value inside a single field is a literal search.
So I searched for `//Object of affection` using the advanced search, fgs_label_s field which leads to this being displayed in the query block: `fgs_label_s:(//Object of desire)` but 0 results.
Solr index contains  
```XML
<str name="fgs_label_s">//Object of desire</str>
```` 
Ok, so i said. Maybe i'm wrong, maybe that is not the expected behavior. I double checked and did a simple search usign `"//Object of affection"` and i got one result as expected.
So i did the same using advanced search... but wait.
My field value was replaced by: `"//Object of affection\"` but query says
````
(-) fgs_label_s:("//Object of affection")
````

same as breadcrumb: 
````
Home » "//Object of affection" (x)
````

but no results! And facet solr pages gives 0 results for any of this combinations (uses filters instead of direct q search right?)

So. I, nor you, should be testing all possible escaping scenarios. But we need to make sure that people can find stuff and consistency. 

So if you have a quick fix to this, that does not break with any other character like `:` then I say let's try it but since i have more than 20 pulls for codefreeze today i would please ask you to do as much as possible extensive testing. 

Another option: not as fancy as regex. But if you need people to be able to search for special characters why not use this idea: escape all double quotes, then surround the result in double quotes. For sure Solr will like that.

adam-vessey committed:
Fix up where multiple tokens are present in the string.

The manner in which it concatenates and invokes stripslashes() would lead
to running stripslashes() many times on the first tokens, but never on the
last. Now, it should only run it once on each.

willtp87 commented:
@DiegoPino The last commit should take care of the erroneous escaping on the final `"//Object of affection\"` .

adam-vessey committed:
Get rid of unnecessary escape of double-quote in regex.

adam-vessey commented:
@DiegoPino: 

> Well, I assumed advanced search considers that the whole value inside a single field is a literal search.

Bad assumption: Spaces in the value would mean that it's searching for `//Object`, `of` and `affection` as three separate terms (both for advanced search _and_ the facet pages search). Searching on strings is not a common behaviour; typically, things are tokenized in some manner. To fit this assumption, we would have to change how it submits the query, surrounding the input with double-quotes based on either some new Islandora Solr config or sniffing the schema from Solr... Not really related to this code; however, to force queries to be submitted as literals, you could add space to the regex to be escaped. 

For double-quotes, they're escaped by design (see the `"` in the regex, formerly `\"` (which worked fine, but was needlessly escaped)), to permit queries with a single double-quotes to be submitted without error. As a point of configuration, this is working as designed.

adam-vessey commented:
... Think I've spotted another issue, with the display of facets...

DiegoPino commented:
@adam-vessey ok. Ping me for that. If possible into this same pull request. Thanks

adam-vessey commented:
@DiegoPino: Yeah, it exists, but it's not really related to this ticket/PR; I'll roll another ticket.

whikloj commented:
Coming in late, I used an old vagrant box with refreshed 7.x code and then @adam-vessey's PR pull. 

Escaping OFF means I did **not** check the new "_Escape Lucene special characters_" checkbox.

So far my test results are:

Search field was `fgs_label_ms`

**Before PR**

| Term |  Escaping OFF | Escaping ON |
| --- | --- | --- |
| `\superCool` | hit, breadcrumb :+1:, search term box ends up with (`superCool`) | perfect |
| `What/ the \heck` | perfect, breadcrumb :+1:, search term box ends up with (`What/ the heck`) |  0 hits, good breadcrumb |
| `B\\To the\\\E` | hit, breadcrumb :+1:, search term box ends up with (`B\To the\E`) |  0 hits, good breadcrumb |
| `Come //on to the \\\/ other side` | hit, breadcrumb :+1:, search term box ends up with (`Come //on to the \/ other side`) |  0 hits, good breadcrumb |
| `//Object of affection\` | hit, :+1: breadcrumb, search term box ends up with (`//Object of affection`) |   0 hits, good breadcrumb |

**With PR**

| Term | Escaping OFF | Escaping On |
| --- | --- | --- |
| `\superCool` | perfect | perfect | 
| `What/ the \heck` | perfect | 0 hits, perfect otherwise |
| `B\\To the\\\E` | perfect | 0 hits, perfect otherwise |
| `Come //on to the \\\/ other side` | perfect | 0 hits, perfect otherwise |
| `//Object of affection\` | perfect | 0 hits, perfect otherwise |

 I'm currently building a new vagrant in-case there are any left over configuration issues. 

I guess I'm probably still not clear on the use case for having these weird characters in the titles, etc. But leaving the checkbox un-checked seems to give the best performance.

willtp87 commented:
@whikloj I think your notes about 0 results are addressed by @adam-vessey's [explanation](https://github.com/Islandora/islandora_solr_search/pull/312#issuecomment-283424351).

adam-vessey commented:
@DiegoPino: Other ticket spawned: https://jira.duraspace.org/browse/ISLANDORA-1925

whikloj commented:
Thanks @willtp87, as I said I may not fully understand the use case. I'll try some different queries to see if I can grok it.

DiegoPino commented:
@adam-vessey @willtp87 quick question about that 0 results explanation. Is the behavior the same with escaping and not escaping? Means 0 on both cases? if that is so, i'm ok with discussing what is expected and not in another issue and go forward with this one. 

Users are not aware of what an admin has or nor setup for them. They expect consistency hopefully across islandora installations. And I guess admins also. Lastly, this whole regex change could need some type of hook that updates the regex. I would say, let's neglect that now since i trust not much people have come into this error or even installed this intermediate versions. But it is breaking and i would expect this will require support.
Will give https://jira.duraspace.org/browse/ISLANDORA-1925 a look now.

whikloj commented:
@willtp87 @adam-vessey I think I get what you are saying about the tokenization of strings.

So I added a test of using `fgs_label_t` versus `fgs_label_s` here are my results all using this PR.

| Term | Escaping |  fgs_label_s | fgs_label_t |
| --- | --- | --- | --- |
| //Object of affection\ | Off | hit | hit |
| //Object | Off | no hit | hit |
| affection\ | Off | no hit | hit |
| //Object of affection\ | On | no hit | hit |
| //Object | On | no hit | hit |
| affection\ | On | no hit | hit |

This seems to be consistent with all the weird strings above.

So with the escaping turned on, you do lose the match on the string field. Which is odd...

But maybe just adding to the description of that checkbox so people are aware that they should only use tokenized fields when that feature is enabled.

DiegoPino commented:
@adam-vessey @willtp87, @whikloj @salomonic "solution" seems to be the quickest way to warn people about (un)expected functionality changes,  so if you could add a simple note to the config field we should be able to move forward and merge.

adam-vessey commented:
@DiegoPino: Not presently, no. With the configurable escaping enabled, spaces are not escaped by default: Note the differences between [the default configurable regex](https://github.com/Islandora/islandora_solr_search/pull/312/files#diff-d33ac2d48f9f2fe37ed473da5e988148R12), and that used by [`islandora_solr_lesser_escape()`](https://github.com/Islandora/islandora_solr_search/blob/fe79ee6ab8e6087380731225493bb3e55d8df92d/includes/utilities.inc#L164).

If we _want_ to change to be more inline with `islandora_solr_lesser_escape()` by default by adding the space, then we could. Shall I?

DiegoPino commented:
@adam-vessey if you ask me, I would say, yeah, go for that missing `| |` 👍 
But i respect any option that leads to consistency, inclusive notes by the field. The good thing about adding space escaping is that basically if you don't want it because you are an advanced user you can remove it. But normal, not regex aware users will love the default.

adam-vessey committed:
Add space to default configurable regex.

DiegoPino commented:
@adam-vessey will have to merge later tonight. Travis was stalled, no logs and probably dead. Restarted the job but I see a lot of issues with Travis not working as expected. Thanks for the hard work here

DiegoPino commented:
Finally! All good with this, making our release branch.
